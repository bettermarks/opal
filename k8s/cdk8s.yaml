language: typescript
app: ts-node main.ts
imports:
  - k8s@1.27.0
