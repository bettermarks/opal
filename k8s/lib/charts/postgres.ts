import { Chart, ChartProps } from "cdk8s";
import {
  ConfigMap,
  ImagePullPolicy,
  ISecret,
  Secret,
  ServiceType,
} from "cdk8s-plus-27";
import { Construct } from "constructs";
import { Namespace } from "../types";
import { IntOrString, KubeDeployment, KubeService } from "../../imports/k8s";

interface PostgresChartProps extends ChartProps {
  image: string;
  name: string;
}

export class PostgresChart extends Chart {
  /**
   * Secret with Postgres credentials
   *
   * keys:
   * - DB_HOST
   * - DB_PORT
   * - DB_USER
   * - DB_PASSWORD
   * - DB_NAME
   * - POSTGRES_PASSWORD
   */
  readonly secret: ISecret;

  constructor(scope: Construct, id: string, props: PostgresChartProps) {
    super(scope, id, props);
    const { name, namespace = Namespace.LICENSING } = props;
    const pgSecret = new Secret(this, "secret", {
      metadata: {
        name: name,
        namespace,
      },
      stringData: {
        DB_PORT: "5432",
        DB_USER: "postgres",
        DB_PASSWORD: "postgres",
        DB_NAME: "postgres",
        POSTGRES_PASSWORD: "postgres",
        POSTGRES_USER: "postgres",
        POSTGRES_DB: "postgres",
      },
    });

    const configMap = new ConfigMap(this, "initdb", {
      metadata: {
        name: `${name}-init-db-script`,
        namespace,
      },
      data: {}, // if you need any init-command put it here...
    });

    new KubeDeployment(this, "pg-deployment", {
      metadata: {
        name,
        namespace,
        labels: {
          app: name,
        },
      },
      spec: {
        selector: {
          matchLabels: {
            app: name,
          },
        },
        replicas: 1,
        template: {
          metadata: {
            labels: {
              app: name,
            },
          },
          spec: {
            containers: [
              {
                name,
                image: props.image,
                imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
                ports: [{ containerPort: 5432 }],
                envFrom: [
                  {
                    secretRef: { name: pgSecret.name },
                  },
                ],
                readinessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                livenessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                volumeMounts: [
                  {
                    name: `${configMap.name}-volume`,
                    mountPath: "/docker-entrypoint-initdb.d/",
                  },
                ],
              },
            ],
            volumes: [
              {
                name: `${configMap.name}-volume`,
                configMap: {
                  name: configMap.name,
                },
              },
            ],
          },
        },
      },
    });

    const pgService = new KubeService(this, id, {
      metadata: {
        name,
        namespace,
      },
      spec: {
        type: ServiceType.CLUSTER_IP,
        ports: [
          {
            port: 5432,
          },
        ],
        selector: {
          app: name,
        },
      },
    });

    pgSecret.addStringData("DB_HOST", pgService.name);
    pgSecret.addStringData("DB_PORT", "5432");
    this.secret = pgSecret;
  }
}
