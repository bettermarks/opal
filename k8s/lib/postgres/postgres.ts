import { IntOrString, KubeDeployment, KubeService } from "../../imports/k8s";
import { LICENSING_SECRET } from "../constants";
import { Namespace } from "../types";
import { Chart, ChartProps } from "cdk8s";
import { ConfigMap, ImagePullPolicy, ServiceType } from "cdk8s-plus-29";
import { Construct } from "constructs";

interface PostgresChartProps extends ChartProps {
  image: string;
  name: string;
}

export class PostgresChart extends Chart {
  constructor(scope: Construct, id: string, props: PostgresChartProps) {
    super(scope, id, props);
    const { name, namespace = Namespace.LICENSING } = props;

    const configMap = new ConfigMap(this, "initdb", {
      metadata: {
        name: `${name}-init-db-script`,
        namespace,
      },
      data: {}, // if you need any init-command put it here...
    });

    new KubeDeployment(this, "pg-deployment", {
      metadata: {
        name,
        namespace,
        labels: {
          app: name,
        },
      },
      spec: {
        selector: {
          matchLabels: {
            app: name,
          },
        },
        replicas: 1,
        template: {
          metadata: {
            labels: {
              app: name,
            },
          },
          spec: {
            containers: [
              {
                name,
                image: props.image,
                imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
                ports: [{ containerPort: 5432 }],
                envFrom: [
                  {
                    secretRef: { name: LICENSING_SECRET },
                  },
                ],
                readinessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                livenessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                volumeMounts: [
                  {
                    name: `${configMap.name}-volume`,
                    mountPath: "/docker-entrypoint-initdb.d/",
                  },
                ],
              },
            ],
            volumes: [
              {
                name: `${configMap.name}-volume`,
                configMap: {
                  name: configMap.name,
                },
              },
            ],
          },
        },
      },
    });

    new KubeService(this, id, {
      metadata: {
        name,
        namespace,
      },
      spec: {
        type: ServiceType.CLUSTER_IP,
        ports: [
          {
            port: 5432,
          },
        ],
        selector: {
          app: name,
        },
      },
    });
  }
}
