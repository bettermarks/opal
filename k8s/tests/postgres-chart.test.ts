import { PostgresChart } from "../lib/postgres/postgres";
import { Testing } from "cdk8s";

describe("postgres-chart", () => {
  test("should match with snapshot", () => {
    const app = Testing.app();
    const chart = new PostgresChart(app, "test-chart", {
      image: "postgres:14",
      name: "licensing-db",
      namespace: "dummy-namespace",
    });
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
