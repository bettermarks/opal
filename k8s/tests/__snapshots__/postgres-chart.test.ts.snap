// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postgres-chart should match with snapshot 1`] = `
[
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "ConfigMap",
    "metadata": {
      "name": "licensing-db-init-db-script",
      "namespace": "dummy-namespace",
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "labels": {
        "app": "licensing-db",
      },
      "name": "licensing-db",
      "namespace": "dummy-namespace",
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "app": "licensing-db",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "licensing-db",
          },
        },
        "spec": {
          "containers": [
            {
              "envFrom": [
                {
                  "secretRef": {
                    "name": "licensing",
                  },
                },
              ],
              "image": "postgres:14",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": {
                "initialDelaySeconds": 10,
                "tcpSocket": {
                  "port": 5432,
                },
              },
              "name": "licensing-db",
              "ports": [
                {
                  "containerPort": 5432,
                },
              ],
              "readinessProbe": {
                "initialDelaySeconds": 10,
                "tcpSocket": {
                  "port": 5432,
                },
              },
              "volumeMounts": [
                {
                  "mountPath": "/docker-entrypoint-initdb.d/",
                  "name": "licensing-db-init-db-script-volume",
                },
              ],
            },
          ],
          "volumes": [
            {
              "configMap": {
                "name": "licensing-db-init-db-script",
              },
              "name": "licensing-db-init-db-script-volume",
            },
          ],
        },
      },
    },
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "licensing-db",
      "namespace": "dummy-namespace",
    },
    "spec": {
      "ports": [
        {
          "port": 5432,
        },
      ],
      "selector": {
        "app": "licensing-db",
      },
      "type": "ClusterIP",
    },
  },
]
`;
